core/drivers/riscv_zkr_rng.c

```diff
@@ -4,46 +4,65 @@
  */
 
 #include <crypto/crypto.h>
+#include <encoding.h>
 #include <kernel/panic.h>
 #include <riscv.h>
 #include <rng_support.h>
 #include <tee/tee_cryp_utl.h>
 
-#define OPST_BIST 0b00
-#define OPST_WAIT 0b01
-#define OPST_ES16 0b10
-#define OPST_DEAD 0b11
-
-TEE_Result hw_get_random_bytes(void *buf, size_t len)
+static inline bool __must_check seed_get_random_u16(uint16_t *val)
 {
-	uint8_t *ptr = buf;
-	uint32_t val = 0;
+	unsigned int retry_count = 0;
+	uint32_t seed = 0;
+	uint32_t opst = 0;
 
-	while (len > 0) {
+	do {
 		/*
-		 * The seed register must be accessed using CSR read-write
-		 * instructions. The write operation is ignored and serves
-		 * to indicate polling and flushing.
+		 * The seed register must be accessed using CSR
+		 * read-write instructions. The write operation
+		 * is ignored and serves to indicate polling and
+		 * flushing.
 		 */
-		val = swap_csr(CSR_SEED, val);
+		seed = swap_csr(CSR_SEED, 0);
+		opst = seed & SEED_OPST;
 
-		switch (val >> 30) {
-		case OPST_BIST:
-		case OPST_WAIT:
-			continue;
-		case OPST_ES16:
-			*ptr++ = val & 0xff;
-			len--;
-			if (len > 0) {
-				*ptr++ = val >> 8;
-				len--;
-			}
-			break;
-		case OPST_DEAD:
+		switch (opst) {
+		case SEED_OPST_ES16:
+			*val = seed & SEED_ENTROPY;
+			return true;
+		case SEED_OPST_DEAD:
 			/* Unrecoverable self-test error */
-			return TEE_ERROR_BAD_STATE;
+			return false;
+		case SEED_OPST_BIST:
+		case SEED_OPST_WAIT:
 		default:
-			break; /* can't happen */
+			retry_count++;
+			continue;
+		}
+	} while (retry_count <= CFG_SEED_RETRY_COUNT_MAX);
```

Comment:
Hi @GseoC,

Thank you for your insights. While a timeout might still fall short in generating sufficient entropy, in that case, increasing RNG_TIMEOUT_US is still needed, right? I would like to keep using retry counter and allow platforms to decide `CFG_SEED_RETRY_COUNT_MAX` based on their use case and quality of entropy source. 

I agreed that the patch still has a bandwidth issue and should add a `cpu_relax()` before each retry iteration, , similar to the Linux kernel's approach:  https://github.com/torvalds/linux/blob/master/arch/riscv/include/asm/archrandom.h#L44

Reply:
Using a timeout value makes more sense to me than a retry count since the process of gathering entropy most likely is timebound. I guess the purpose of the timeout is to catch the unlikely case where we can't get any random number and need a way to abort the operation. Depending on where this happens it will most likely mean panicking a TA. So the number should be large enough that it can't happen during normal operations. Trying with different values until it doesn't occur seems like a fragile approach.

Reply:
Sure, I will switch to the timeout method. Since there is not actual hardware I can test with, I will assume a 1-second fixed timeout. We may still need something like `CFG_SEED_TIMEOUT` for platforms with varying entropy bandwidth.

Reply:
> Could be yes, but the hardware block may have a duration (dependent of its clock) above which the entropy should have been generated in the worst case scenario. Is it the case? If you use a timeout, I suggest IO_READ32_POLL_TIMEOUT().

Thank you for the suggestion! However, in this case, the seed is a register, so it can't be accessed with `IO_READ32_POLL_TIMEOUT()`